Fix GDD under fastpath.

kainwen-zz opened the pull request:
With 9.2 merge fastpath is add, GDD is disable temporarily as it cannot properly
recognize all locks with fastpath.

In the function set the HoldTillEndXact field, locallock->lock might be NULL if
the lock is held via fast path.

In this commit, I add a field fpHoldTillEndXact(like fpRelId) in PROC struct to save this
info.

kainwen commented:
@guofengrichard 
I have tested it using isolation2/gdd cases. It passes.

I have no idea whether we need to add more cases to test fast path.

Please review it, thanks.

remotefish commented on src/include/storage/proc.h:
```diff
@@ -186,6 +186,7 @@ struct PGPROC
 	/* Lock manager data, recording fast-path locks taken by this backend. */
 	uint64		fpLockBits;		/* lock modes held for each fast-path slot */
 	Oid			fpRelId[FP_LOCK_SLOTS_PER_BACKEND];		/* slots for rel oids */
+	bool		fpHoldTillEndXact[FP_LOCK_SLOTS_PER_BACKEND];		/* slots for gdd flags */
```
As `FP_LOCK_SLOTS_PER_BACKEND` is 16, I wonder whether should we define `fpHoldTillEndXact` as a `uint16` instead of an array of bool to save spaces in this struct.

kainwen-zz replied:
bool is implemented as char. I think char is OK? @pivotal-ning-yu

remotefish replied:
Using bool or char will add 16 bytes to PGPROC struct, that's too much. Why is `fpLockBits` defined as `uint64` instead of `uint8 [16]`? The same purpose.

kainwen-zz replied:
@pivotal-ning-yu 
You mean using a integer's bit to mimic array?

It is tricky because it depends on FP_LOCK_SLOTS_PER_BACKEND.

remotefish replied:
fast path implementation is already playing this trick as `fpLockBits`. Wasting memory in `PGPROC` struct should always be avoided.

guofengrichard replied:
`FP_LOCK_SLOTS_PER_BACKEND` is not enough for your purpose with `fpHoldTillEndXact`.  Note that one slot here can stand for at most `FAST_PATH_BITS_PER_SLOT` locks.

remotefish replied:
@guofengrichard each slot can contain multiple lock modes, but not multiple locks. Only lock modes in [1, 3] are tracked via fast path.

guofengrichard replied:
Yes, that is what `FAST_PATH_BITS_PER_SLOT` is used for. Is the `holdTillEndXact` info per lockmode or per lock?

remotefish commented on src/backend/storage/lmgr/lock.c:
```diff
@@ -2059,8 +2060,12 @@ LockSetHoldTillEndXact(const LOCKTAG *locktag)
 	}
+
+	relid = locktag->locktag_field2;
+	setFPHoldTillEndXact(relid);
```
Looks like it's better call this fastpath one first, depends on its return value (we should let this function return a bool to identify whether the lock is in fastpath) we could skip the slow for loop.

kainwen-zz replied:
@pivotal-ning-yu  I am not sure whether we could skip the for loop even if we have found some locks in fast path array.

Any ideas @guofengrichard

remotefish replied:
@kainwen gdd logic is that once a relation has unbalanced ref counts we mark all its lock modes as unreleasable, for fast path could we assume that a lock is either tracked in fast path or in normal local hash but never both? If this assumption is true then we sure could skip the full loop. You could add some check or at least an assert here to verify this assumption.

guofengrichard replied:
That's right. A relation lock would exist either in fast-pach or in shared lock table. You can refer to the logic `LockRelease()`.

kainwen-zz replied:
@guofengrichard 
Good news. I will improve the code.

guofengrichard replied:
The logic in `LockRelease()` is first attempt fast release if the lock is eligible for the fast path. This attempt is not guaranteed to be always successful, because other backends may move the lock to hash table, or the lock is not entered in fast-path due to no available slot. 
I think that logic also applies in this function.

remotefish commented on src/backend/storage/lmgr/lock.c:
```diff
@@ -377,7 +377,7 @@ static void LockRefindAndRelease(LockMethod lockMethodTable, PGPROC *proc,
 					 bool decrement_strong_lock_count);
 
-
+static void setFPHoldTillEndXact(Oid relid);
```
Please move this declaration one line above to sit together with other function declarations, and keep the two existing blank lines here.

kainwen-zz replied:
Thanks. Will improve the code style here.

remotefish commented on src/backend/storage/lmgr/lock.c:
```diff
@@ -4264,3 +4270,25 @@ VirtualXactLock(VirtualTransactionId vxid, bool wait)
+			proc->fpRelId[f] != relid)
+			continue;
+
+		proc->fpHoldTillEndXact[f] = true;
```
If we have found the lock in fastpath could we break now?

kainwen-zz replied:
I have considered this. But I am not sure whether the remaining array contains a relation lock for the same relid. @guofengrichard What is your idea?

guofengrichard replied:
A `break` here is good since one relid only occupies one slot.
In addition, suggest to use the following pattern to look for the matched slot, as most time there is no necessary to get the lock bits.
```
 if (relid != proc->fpRelId[f] || FAST_PATH_GET_BITS(proc, f) == 0)
                continue;
```

kainwen-zz replied:
@guofengrichard Thanks. I will improve the code.

remotefish commented on src/backend/storage/lmgr/lock.c:
```diff
@@ -4264,3 +4270,25 @@ VirtualXactLock(VirtualTransactionId vxid, bool wait)
 }
+
+static void
+setFPHoldTillEndXact(Oid relid)
```
please add function comments.

kainwen-zz replied:
Ok. Will add some comments.

guofengrichard commented on src/backend/storage/lmgr/lock.c:
```diff
@@ -3545,7 +3552,7 @@ GetLockStatusData(void)
 								   proc->gxact.gxid :
 								   proc->localDistribXactData.distribXid;
-			instance->holdTillEndXact = false;
+			instance->holdTillEndXact = proc->fpHoldTillEndXact[f];
```
I don't think the logic is right here. One `instance` for fast-path stands for at most 3 locks. So which lock `holdTillEndXact` is describing for?

kainwen-zz replied:
@guofengrichard I understand slots wrong.

kainwen-zz replied:
@guofengrichard 

As you mentioned:
```
One instance for fast-path stands for at most 3 locks. So which lock holdTillEndXact is describing for?
```

And instance->holdTillEndXact is a bool. How can a bool describe 3 locks' status?

guofengrichard commented:
Locks in per-backend fast-path arrays may be transferred to the shared hash table at runtime. When that happens,  the `HoldTillEndXact` info in fast-path should be passed to hash table.

kainwen commented:
@guofengrichard Thanks. I will refactor this PR.

kainwen committed:
Fix GDD under fastpath.

With 9.2 merge fastpath is add, GDD is disable temporarily as it cannot properly
recognize all locks with fastpath.

In the function set the HoldTillEndXact field, locallock->lock might be NULL if
the lock is held via fast path.

In this commit, I add a field fpHoldTillEndXactBits(like fpLockBits) in PROC
struct to save this info.

kainwen commented:
@guofengrichard @pivotal-ning-yu 
I have refactored the code, please help review it.

I tested it via GDD cases and passed. But I have no idea if we should or how to add some caese
for fastpath situation.

guofengrichard commented on src/backend/storage/lmgr/lock.c:
```diff
@@ -2053,13 +2072,8 @@ LockSetHoldTillEndXact(const LOCKTAG *locktag)
-		 * This code should be revised to make GDD work under fast path.
-		 */
-//		locallock->lock->holdTillEndXact = true;
+		if (locallock->lock)
```
An assertion is better here.

kainwen-zz replied:
@guofengrichard 
try_heap_open(, NoLock, );
...
relation_close(r, Nolock); --> will leave locallock->lock is NULL in this branch.

However, in master branch of gpdb(not merge branch), locallock is NULL so we never execute the assignment.

I think we could keep it like this.

And refactor it in another MPP story.

@pivotal-ning-yu

guofengrichard reviewed (approved):
LGTM
